querydsl-core/src/main/java/com/querydsl/core/support/SerializerBase.java

```diff
@@ -46,8 +58,9 @@
 
     private String anonParamPrefix = "_";
 
-    private Map<Object, String> constantToNamedLabel;
-    private Map<Object, Integer> constantToNumberedLabel;
+    protected final List<Object> constants = new LinkedList<>();
```

Comment:
Is there a need for this? Seems like the only difference to `constantToLabel` keys is order, is that relevant?
Seems like it should be possible to use a single data structure instead of keeping two in sync.

Reply:
Unfortunately it is not, and that is because `constantToLabel` works very differently across the various serializers for different dialects.

Constant to label is a mapping used primarily by JDO and CollQuerySerializer to create a unique label for unique objects and store them as a variable in the context of the query (so this is in fact not really a constant or a parameter at all).

As query parameters objects may occur different times using various mappings. Therefore the SQL serializer already maintained a list of constants instead of relying on constantToLabel. The proper fix was to use this for JPQL as well.

I did try to combine the two datastructures and it is possible as long as all labels are unique and the mapping is inverted to labelToConstant. However, this approach makes the implementation in SQL and JPQL overly complex and error prone (as this issues already show) and it also  defeats optimizations  that were deliberately made for JDO and CollQuerySerializer. In the end I just stumbled upon lots of overriden methods and serializer hacks to get all tests passing in all modules.

I am also worried that changing `constantToLabel` to `labelToConstant` is going to be a massive pain in the ass for integrators that implement different serializers based on this API or even use `getLabelToConstant()` for logging / parameter binding purposes  in their own applications.

This approach is simpler, more generic and only changes recently updated API's while leaving the API's that have been there for years in tact.